### PR TITLE
Mobile: display filmstrip on the short side

### PR DESCRIPTION
**Pull request by paweldomas**
Adds the ability to detect app area's aspect ratio on react-native
through the features/base/aspect-ratio.

Makes conference, filmstrip and toolbox react to the aspect ratio
changes and display filmstrip on the shorter side of the screen.

<hr/>

img_0003
img_0004

### CODE REVIEW

**Comment by saghul** [comment]
General comment. I apologize in advance, you are going to hate me @paweldomas, but I have to suggest a new name for this "aspect ratio" stuff. An aspect ratio should be, well, a ratio. As in 4/3 or 16/9, not "narrow" or "wide". I think using "orientation" here is a better fit, having it be "landscape" or "portrait". WDYT?

**Comment by saghul** [comment]
Another general comment: while this `onLayout` approach is a neat simple trick, I wonder if we'll run into the problem listed in the limitations section here: https://corbt.com/posts/2016/03/16/detecting-orientation-in-react-native.html

File: react/features/toolbox/components/Toolbox.native.js [hunk 1]
```diff
@@ -119,15 +120,25 @@ class Toolbox extends Component {
      * @returns {ReactElement}
      */
     render() {
+        if (!this.props._visible) {
```
Comment (saghul) [inline react/features/toolbox/components/Toolbox.native.js]: why is this necessary? this is already implemented at the container level, isn't it?

File: react/features/toolbox/components/Toolbox.native.js [hunk 2]
```diff
@@ -119,15 +120,25 @@ class Toolbox extends Component {
...
+                        : styles.toolbarContainerWide } >
                 {
-                    this._renderPrimaryToolbar()
+                    isNarrowAspectRatio(this)
+                        ? this._renderSecondaryToolbar()
+                        : this._renderPrimaryToolbar()
                 }
                 {
-                    this._renderSecondaryToolbar()
+                    isNarrowAspectRatio(this)
```
Comment (saghul) [inline react/features/toolbox/components/Toolbox.native.js]: this code looks weird when reading, is there no other way to write it a bit come compact? maybe a helper method called `_renderToolbars` which takes the orientation into consideration?

File: react/features/toolbox/components/styles.js [hunk 3]
```diff
@@ -6,7 +6,7 @@ import { BoxModel, ColorPalette, createStyleSheet } from '../../base/styles';
  * @type {Object}
  */
 const _toolbar = {
-    flex: 1,
+    flex: 0,
```
Comment (saghul) [inline react/features/toolbox/components/styles.js]: is this correct? you are using flex options for primaryToolbar, but flex will be 0

**Comment by saghul** [comment]
Updated the original comment with 2 screenshots. Some observations:

- The portrait oriented one looks really weird to me with the buttons above the filmstrip. I wonder if we can do something else, like put the filmstrip on the upper part, and the secondary toolbar coming from the bottom instead of from the top.
- On the landscape version, I think the local view should be the one on top.

I'm sorry for all these comments. Please let me know if we want to separate the design discussion from this PR.

**Comment by lyubomir** [comment]
> I have to suggest a new name for this "aspect ratio" stuff. An aspect ratio should be, well, a ratio. As in 4/3 or 16/9, not "narrow" or "wide". I think using "orientation" here is a better fit, having it be "landscape" or "portrait". WDYT?

@paweldomas and I explicitly chose a name other than "orientation" because (1) "orientation" gets us into the land of device orientation, landscape, and portrait and we don't really want device orientation. As is apparent from the implementation, (2) we're not even building the feature on the device's orientation. (3) We do want to adapt the user interface depending on the aspect ratio of a view regardless of the device's orientation. With multitasking we can get into aspect ratio's which are "opposite" to the device's orientation.

A weakness of the implementation in general which we're aware of is redux and the fact that there's only one aspect ratio value i.e. only one `AspectRatioDetector`.

File: react/features/toolbox/components/styles.js (hunk 3 above)
Comment (lyubomir) [reply react/features/toolbox/components/styles.js]: https://developer.mozilla.org/en-US/docs/Web/CSS/flex says:

The flex CSS property specifies how a flex item will grow or shrink so as to fit the space available in its flex container. This is a shorthand property that sets flex-grow, flex-shrink, and flex-basis.

**Comment by saghul** [comment]
> Another general comment: while this onLayout approach is a neat simple trick, I wonder if we'll run into the problem listed in the limitations section here: https://corbt.com/posts/2016/03/16/detecting-orientation-in-react-native.html

This doesn't seem to be a problem, I got access to an iPad and tested it. 👍

**Comment by saghul** [comment]
Yet another observation: the top margin in the filmstrip vs toolbar don't seem to match (see 2nd screenshot), but they probably should.

**Comment by saghul** [comment]
> @paweldomas and I explicitly chose a name other than "orientation" because (1) "orientation" gets us into the land of device orientation, landscape, and portrait and we don't really want device orientation. As is apparent from the implementation, (2) we're not even building the feature on the device's orientation. (3) We do want to adapt the user interface depending on the aspect ratio of a view regardless of the device's orientation. With multitasking we can get into aspect ratios which are "opposite" to the device's orientation.

I see. I'd still try to find some wording which doesn't include "ratio" since we have no concept of ratio. View aspect?

> A weakness of the implementation in general which we're aware of is redux and the fact that there's only one aspect ratio value i.e. only one AspectRatioDetector.

If by that you mean the fact that they cannot be nested... I think we can live with that :-)

File: react/features/toolbox/components/Toolbox.native.js (hunk 1 above)
Comment (paweldomas) [reply react/features/toolbox/components/Toolbox.native.js]: it doesn't work as expected - see Container implementation - it barley adjusts width and height which may not always be respected by child components

File: react/features/toolbox/components/Toolbox.native.js (hunk 2 above)
Comment (paweldomas) [reply react/features/toolbox/components/Toolbox.native.js]: can do - yes

File: react/features/toolbox/components/Toolbox.native.js (hunk 1 above)
Comment (saghul) [reply react/features/toolbox/components/Toolbox.native.js]: then we should fix it there IMHO. Note that now that the filmstrip is always visible we can remove the workaround and return null on all platforms.

File: react/features/toolbox/components/styles.js (hunk 3 above)
Comment (saghul) [reply react/features/toolbox/components/styles.js]: But when set to 0 it uses height and width for layout instead of flexbox (according to the React Native docs: https://facebook.github.io/react-native/docs/layout-props.html#flex). My CSS kung-fu is very weak, I just noticed that we are using flex options when flex is kinda disabled and I wondered if that's ok.

**Comment by paweldomas** [comment]
> Yet another observation: the top margin in the filmstrip vs toolbar don't seem to match (see 2nd screenshot), but they probably should.

If I adjust filmstrip's margin then we'll have less space for the thumbnail, as it will also make sense to have the same margin on the bottom.

Another option would be to move the toolbox up, but I don't know if that's a good idea.

File: react/features/toolbox/components/Toolbox.native.js (hunk 1 above)
Comment (paweldomas) [reply react/features/toolbox/components/Toolbox.native.js]: ok

Comment (paweldomas) [reply react/features/toolbox/components/Toolbox.native.js]: but on the other hand, why even bother with doing the whole render logic if there's nothing to be rendered ?

File: react/features/toolbox/components/styles.js (hunk 3 above)
Comment (paweldomas) [reply react/features/toolbox/components/styles.js]: How I found this working was that the toolbar were adjusting their sizes to match the size of the contents (this style is shared between the primary and the secondary toolbars).

**Comment by paweldomas** [comment]
As part of the force push I've made the margins consistent.

**Comment by paweldomas** [comment]
> The portrait oriented one looks really weird to me with the buttons above the filmstrip. I wonder if we can do something else, like put the filmstrip on the upper part, and the secondary toolbar coming from the bottom instead of from the top.
> On the landscape version, I think the local view should be the one on top.

I see your point and I agree it may look weird especially if there are not many thumbnails, but I'd rather only keep note of those observations for now and make adjustments once more people get to try it and then come up with improvements once we come to some agreement on all those details.

File: react/features/base/aspect-ratio/components/AspectRatioAware.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                 * Either {@link ASPECT_RATIO_NARROW} or
+                 * {@link ASPECT_RATIO_WIDE}.
+                 */
+                aspectRatio: PropTypes.symbol
+            }
+
+            /**
+             * Implement's React render method to wrap the nested component.
+             *
+             * @returns {XML}
```
Comment (saghul) [inline react/features/base/aspect-ratio/components/AspectRatioAware.js]: `{ReactElement}` I think?

File: react/features/base/aspect-ratio/components/AspectRatioDetector.native.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+     * Renders the root view and it's children.
+     *
+     * @returns {Component}
+     */
+    render() {
+        return (
+            <View
+                onLayout = { this.props._onLayout }
+                style = { styles.aspectRatioDetectorStyle } >
+                {this.props.children}
```
Comment (saghul) [inline react/features/base/aspect-ratio/components/AspectRatioDetector.native.js]: style: space after { and before } 

Sidenote: I think eslint doesn't catch these because it's not an object literal.

File: react/features/base/aspect-ratio/actions.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * values.
+ *
+ * @param {number} width - The width of the app's area used on the screen.
+ * @param {number} height - The height of the app's area used on the screen.
+ * @returns {{
+ *      type: SET_ASPECT_RATIO,
+ *      aspectRatio: Symbol
+ * }}
+ */
+export function calculateNewAspectRatio(width: number, height: number): Object {
```
Comment (saghul) [inline react/features/base/aspect-ratio/actions.js]: why not call this `setAspectRatio` and do the calculation outside of this function?

**Comment by paweldomas** [comment]
jenkins test this please
